Customer Segmentation Menggunakan Fuzzy C-Means

1. Import beberapa libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

2. Load data

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Mall_Customers.csv to Mall_Customers (4).csv


In [ ]:
import io
df = pd.read_csv(io.BytesIO(uploaded['Mall_Customers.csv']))
df

CustomerID  Gender  Age  Annual Income (k$)  Spending Score (1-100)
0             1    Male   19                  15                      39
1             2    Male   21                  15                      81
2             3  Female   20                  16                       6
3             4  Female   23                  16                      77
4             5  Female   31                  17                      40
..          ...     ...  ...                 ...                     ...
195         196  Female   35                 120                      79
196         197  Female   45                 126                      28
197         198    Male   32                 126                      74
198         199    Male   32                 137                      18
199         200    Male   30                 137                      83

[200 rows x 5 columns]

3. Preprocessing Data

In [ ]:
from sklearn.preprocessing import StandardScaler, LabelEncoder
df.rename(columns = {'Annual Income (k$)': 'Annual Income','Spending Score (1-100)': 'Spending Score'}, inplace = True)
df.drop('CustomerID', axis = 1, inplace = True)
labeler = LabelEncoder()
df['Gender'] = labeler.fit_transform(df['Gender'])

In [ ]:
scaler = StandardScaler()
df_new = pd.DataFrame(scaler.fit_transform(df), columns = df.columns)

4. Menentukan Jumlah Cluster Menggunakan Silhouette Coefficient

In [ ]:
from sklearn.metrics import silhouette_score

In [ ]:
mall = df_new.to_numpy()

In [ ]:
from fcmeans import FCM
cluster = []
for i in range(2,11):
    fcm = FCM(n_clusters = i)
    fcm.fit(mall)
    fcm_labels = fcm.predict(mall)
    cluster.append(silhouette_score(mall,fcm_labels))
print('Jumlah cluster terbaik :',cluster.index(max(cluster)))

Jumlah cluster terbaik : 6


Dari grafik diatas, maka didapat cluster terbaik adalah berjumlah 6

5. Clustering Menggunakan Fuzzy C-Means

In [ ]:
fcm = FCM(n_clusters=6)
fcm.fit(mall)
fcm_labels = fcm.predict(mall)

In [ ]:
df["Cluster"] = fcm_labels

In [ ]:
df

Gender  Age  Annual Income  Spending Score  Cluster
0         1   19             15              39        1
1         1   21             15              81        1
2         0   20             16               6        3
3         0   23             16              77        3
4         0   31             17              40        3
..      ...  ...            ...             ...      ...
195       0   35            120              79        4
196       0   45            126              28        2
197       1   32            126              74        4
198       1   32            137              18        2
199       1   30            137              83        4

[200 rows x 5 columns]

6. Analisis Cluster

In [ ]:
cluster1 = df[df["Cluster"]==0]
cluster2 = df[df["Cluster"]==1]
cluster3 = df[df["Cluster"]==2]
cluster4 = df[df["Cluster"]==3]
cluster5 = df[df["Cluster"]==4]
cluster6 = df[df["Cluster"]==5]

Cluster pertama berisikan pelanggan perempuan dewasa dalam rentang 34 sampai 68 tahun berpenghasilan 20 ribu sampai 101 ribu dolar dan pengeluaran dalam rentang 5 sampai 59

Cluster 2 berisikan pelanggan pria mulai dari 18 tahun sampai 40 tahun dengan pendapatan 15 ribu sampai 101 ribu dolar dan pengeluaran dalam rentang 35 sampai 97

Cluster 3 berisikan pelanggan mayoritas pria berusia mulai dari 19 tahun sampai 59 tahun dengan penghasilan 71 ribu sampai 137 ribu dolar dengan pengeluran mulai dari 1 sampai 36

Cluster 4 berisikan pelanggan perempuan dengan usia 18 tahun sampai 36 tahun yang berpenghasilan 16 ribu sampai 76 ribu dolar dengan pengeluaran 6 sampai 99

Cluster 5 berisikan pelanggan mayoritas perempuan dengan umur mulai dari 27 sampai 41 tahun berpenghasilan 70 ribu sampai 137 ribu dolar dan pengeluaran kisaran 39 sampai 97

Cluster 6 berisikan pelanggan pria berusia 37 sampai 70 tahun dengan penghasilan 19 ribu sampai 63 ribu dolar dengan pengeluaran 3 sampai 60